# HG changeset patch
# User Felipe G. Nievinski <fgnievinski@gmail.com>
# Date 1392086563 7200
#      Tue Feb 11 00:42:43 2014 -0200
# Node ID 1363d909c5770799775dc33cb71507c09022dd7e
# Parent  b0aba84cf80fb3958d696ee8e685059aeaf3c32c
run.m: Don't undo changes to current dir done by script that was run (bug #41543).

* run.m: After running script, check that current directory was the same one
where script was located.  If it is, cd back to original directory, otherwise
leave user in the directory that the script changed to.

diff --git a/scripts/miscellaneous/run.m b/scripts/miscellaneous/run.m
--- a/scripts/miscellaneous/run.m
+++ b/scripts/miscellaneous/run.m
@@ -26,19 +26,20 @@
 ## typing their name.  For scripts not located on the load path, use @code{run}.
 ##
 ## The file name @var{script} can be a bare, fully qualified, or relative
 ## filename and with or without a file extension.  If no extension is specified,
 ## Octave will first search for a script with the @file{".m"} extension before
 ## falling back to the script name without an extension.
 ##
 ## Implementation Note: If @var{script} includes a path component, then
-## @code{run} first changes the directory to the directory where @var{script}
-## is found.  @code{run} then executes the script, and returns to the original
-## directory. 
+## @code{run} first changes the working directory to the directory where
+## @var{script} is found.  Next, the script is executed.  Finally, @code{run}
+## returns to the original working directory unless @code{script} has
+## specifically changed directories.
 ## @seealso{path, addpath, source}
 ## @end deftypefn
 
 function run (script)
 
   if (nargin != 1)
     print_usage ();
   endif
@@ -60,17 +61,19 @@ function run (script)
   if (! isempty (d))
     if (exist (d, "dir"))
       wd = pwd ();
       unwind_protect
         cd (d);
         evalin ("caller", sprintf ('source ("%s%s");', f, ext),
                 "rethrow (lasterror ())");
       unwind_protect_cleanup
-        cd (wd);
+        if (strcmp (d, pwd ()))
+          cd (wd);
+        endif
       end_unwind_protect
     else
       error ("run: the path %s doesn't exist", d);
     endif
   else
     if (! isempty (ext))
       script = which (script);
     else
